rootston/keyboard.c

```diff
@@ -116,86 +155,64 @@ static bool keyboard_keysym_press(struct roots_keyboard *keyboard,
-	uint32_t consumed = xkb_state_key_get_consumed_mods2(
+static size_t keyboard_keysyms_translated(struct roots_keyboard *keyboard,
+		xkb_keycode_t keycode, const xkb_keysym_t **keysyms,
+		uint32_t *modifiers) {
+	*modifiers = wlr_keyboard_get_modifiers(keyboard->device->keyboard);
+	xkb_mod_mask_t consumed = xkb_state_key_get_consumed_mods2(
 		keyboard->device->keyboard->xkb_state, keycode, XKB_CONSUMED_MODE_XKB);
+	*modifiers = *modifiers & ~consumed;
 
-	// TODO: actually use this value
-	modifiers = modifiers & ~consumed;
-
-	bool handled = false;
-	for (int i = 0; i < syms_len; i++) {
-		if (state) {
-			bool keysym_handled =
-				keyboard_keysym_press(keyboard, syms[i]);
-			handled = handled || keysym_handled;
-		} else { // WLR_KEY_RELEASED
-			keyboard_keysym_release(keyboard, syms[i]);
-		}
-	}
-
-	return handled;
+	return xkb_state_key_get_syms(keyboard->device->keyboard->xkb_state,
+		keycode, keysyms);
 }
+
 /*
- * Process keypresses from the keyboard as if modifiers didn't change keysyms.
+ * Get keysyms and modifiers from the keyboard as if modifiers didn't change
+ * keysyms.
  *
  * This avoids the xkb keysym translation based on modifiers considered pressed
- * in the state and uses the list of modifiers saved on the rootston side.
+ * in the state.
  *
- * This will trigger the keybind: [Alt]+[Shift]+2
+ * This will trigger the keybind: Alt + Shift + 2
  */
-static bool keyboard_keysyms_simple(struct roots_keyboard *keyboard,
-		uint32_t keycode, enum wlr_key_state state) {
-	const xkb_keysym_t *syms;
+static size_t keyboard_keysyms_raw(struct roots_keyboard *keyboard,
+		xkb_keycode_t keycode, const xkb_keysym_t **keysyms,
+		uint32_t *modifiers) {
+	*modifiers = wlr_keyboard_get_modifiers(keyboard->device->keyboard);
+
 	xkb_layout_index_t layout_index = xkb_state_key_get_layout(
 		keyboard->device->keyboard->xkb_state, keycode);
-	int syms_len = xkb_keymap_key_get_syms_by_level(
-		keyboard->device->keyboard->keymap, keycode, layout_index, 0, &syms);
-
-	bool handled = false;
-	for (int i = 0; i < syms_len; i++) {
-		if (state) {
-			bool keysym_handled = keyboard_keysym_press(keyboard, syms[i]);
-			handled = handled || keysym_handled;
-		} else { // WLR_KEY_RELEASED
-			keyboard_keysym_release(keyboard, syms[i]);
-		}
-	}
-
-	return handled;
+	return xkb_keymap_key_get_syms_by_level(keyboard->device->keyboard->keymap,
+		keycode, layout_index, 0, keysyms);
 }
 
 void roots_keyboard_handle_key(struct roots_keyboard *keyboard,
 		struct wlr_event_keyboard_key *event) {
-	uint32_t keycode = event->keycode + 8;
-
-	bool handled = keyboard_keysyms_xkb(keyboard, keycode, event->state);
+	xkb_keycode_t keycode = event->keycode + 8;
 
+	uint32_t modifiers;
+	const xkb_keysym_t *keysyms;
+	size_t keysyms_len = keyboard_keysyms_translated(keyboard, keycode,
+		&keysyms, &modifiers);
+	bool handled = keyboard_handle_keysyms(keyboard,
+		keyboard->pressed_keysyms_translated, modifiers, keysyms, keysyms_len,
+		event->state);
```

Comment:
There is a subtle bug here with this approach. If I bind to `a` and `b` and then press `a`, press `b`, release `a`, release `b`, then the `a` binding will run twice, once when I press `a` and once when I press `b` when I would expect it to only run each binding once.

Reply:
I think the easy fix to this is to make sure the keysyms pressed are *exactly* the ones specified by the binding.